app/api/src/crud/crud_read_heatmap.py

```diff
@@ -470,33 +467,121 @@ def quantile_classify(self, calculations):
-        directory = os.path.join(base_path, str(study_area_id), "h3")
-        file_name = os.path.join(directory, f"{resolution}.geojson")
-        dataframe = gpd.read_file(file_name)
-        dataframe["bulk_id"] = dataframe["bulk_id"].apply(int, base=16)
-        return dataframe
+            parent_tags[key] = tags(unique[0])
+
+        return parent_tags
 
-    def merge_frames(self, hexagons: pd.DataFrame, calculations_dataframes: list[pd.DataFrame]):
-        pass
+    def create_grids_unordered_map(self, grids: dict):
+        """
+        Convert grids to unordered map for fast lookup
+        """
+        indexes = range(grids.size)
+        grids_unordered_map = dict(zip(grids, indexes))
+        return grids_unordered_map
+
+    def get_calculations_null_array(self, grids):
+        return np.full(grids.size, np.nan, np.float32)
+
+    def create_grid_pointers(self, grids_unordered_map, parent_tags):
+        grid_pointers = {}
+        get_id = lambda tag: grids_unordered_map.get(tag, -1)
+        for key, parent_tag in parent_tags.items():
+            if not parent_tag.size:
+                grid_pointers[key] = parent_tag.copy()
+                continue
+            grid_pointers[key] = np.vectorize(get_id)(parent_tag)
+        return grid_pointers
+
+    def create_calculation_arrays(self, grids, grid_pointers, calculations):
+        calculation_arrays = {}
+        for key, grid_pointer in grid_pointers.items():
+            if not grid_pointer.size:
+                calculation_arrays[key] = grid_pointer.copy()
+                continue
+            calculation_arrays[key] = self.get_calculations_null_array(grids)
+            mask = grid_pointer != -1
+            masked_grid_pointer = grid_pointer[mask]
+            calculation_arrays[key][masked_grid_pointer] = calculations[key][mask]
+        return calculation_arrays
+
+    def create_quantile_arrays(self, calculations):
+        quantile_arrays = {}
+        for key, calculation in calculations.items():
+            quantile_arrays[key] = heatmap_core.quantile_classify(calculation)
+        return quantile_arrays
+
+    def reorder_calculations(self, calculations, grids, uniques):
+        """
+        Reorder calculations to match the order of grids
+        """
+        # convert grid_id from int to hex
+        sample_grid_id = h3.h3_to_string(grids[0])
+        target_resolution = h3.h3_get_resolution(sample_grid_id)
+        uniques_parent_tags = self.tag_uniques_by_parent(uniques, target_resolution)
+        grids_unordered_map = self.create_grids_unordered_map(grids)
+        grid_pointer = self.create_grid_pointers(grids_unordered_map, uniques_parent_tags)
+        calculations = self.create_calculation_arrays(grids, grid_pointer, calculations)
+        quantiles = self.create_quantile_arrays(calculations)
+        return calculations, quantiles
+
+    def convert_parent_tags_to_geojson(self, parent_tags: dict):
+        tags = set()
```

Comment:
Is this function used or only for testing?

Reply:
Yes. It is just for testing.